### PR TITLE
[Bindless][Exp] Windows & DX12 interop. Semaphore ops can take values.

### PR DESCRIPTION
**Pull request by przemektmalon**
The following PI functions have been added to support importing various external handle types for both memory and semaphores:
  - `piextImportExternalMemory`
  - `piextImportExternalSemaphore`

The following PI functions and their corresponding `pi2ur` functions are now deprecated:
  - `piextImportExternalSemaphoreOpaqueFD`
  - `piextMemImportOpaqueFD`

All Vulkan tests have been updated to work on both Windows and Linux. Comments have been added to the Vulkan test files to make it easier to read and understand the code.

Support has been added for interoperability of certain DirectX 12 resources, namely dedicated memory resources and fences. A test has been added that uses both functionalities.

Support has been added for semaphore operations to take values passed by the user. The semaphore will either signal a given value, or wait for a value of the user's choice.

### CODE REVIEW

**Comment by przemektmalon** [comment]
Associated UR MR: https://github.com/oneapi-src/unified-runtime/pull/1639

File: sycl/plugins/unified_runtime/pi2ur.hpp
Comment (JackAKirk) [inline sycl/plugins/unified_runtime/pi2ur.hpp:5512]: I don't see any functional reason to add the sycl context as an argument here? All that you are doing with it is checking that it casting it to a ur context matches the ur context of the ur device later on in `urBindlessImagesImportExternalSemaphoreExp`, but you get the cuContext to set from the device anyway.

If arguments don't have a purpose, then they should not be added to a new interface. This leads to unnecessary problems, unclear documentation, and excess code to maintain with no purpose.

Comment (przemektmalon) [reply sycl/plugins/unified_runtime/pi2ur.hpp:5512]: You're right that the CUDA adapter does not need both the context and device parameters, as the `ur_device_handle_t` for the CUDA adapter has a `getContext` method. However, not all adapters implement the device struct in the same way. For example, the LevelZero adapter's `ur_device_handle_t` does not have a way of retrieving a context from the device. Other adapters may also differ.

Since I'm not sure what the implementation of this functionality will look like in every adapter, I think it better to play it safe and provide both the context and device parameters. Moreover, virtually every other PI/UR function that takes a `device` parameter, also takes a `context` parameter. So while it might not be necessary for the CUDA backend, I can only make the assumption that this is not the case for every backend. Having both parameters present from the start will prevent having to break ABI in the future for backends that may need both.

Am I missing something? Is there no good reason that almost every PI/UR function pairs the `device` parameter with a `context`?

Comment (JackAKirk) [reply sycl/plugins/unified_runtime/pi2ur.hpp:5512]: > plementation of this functionality will look like in every adapter, I think it better to play it safe and provide both the context and device parameters. Moreover, virtually every other PI/UR function that takes a `device` parameter, also takes a `context` parameter. So while it might not be necessary for the CUDA backend, I can only make the assumption that this is not the case for every backend. Having both parameters present from the start will prevent having to break ABI in the future for backends that may need both.

the correspondence between cuContext and level zero equivalent is a ZeDevice: Level zero also does not need any concept of context for any kind of USM based operations. They may choose to take context arguments and enforce constraints like you can't do usm copies between devices in different contexts. But there is no practical reason to do this.

I've tried to explain why other backends should also never need a context concept: the main point is point 1) here:
https://github.com/oneapi-src/unified-runtime/pull/1639#issuecomment-2122316380

I don't have much hope of persuading you, but it would be irresponsible for me to point this out in a review. Adding these redundant arguments to new functions does lead to problems down the line. And generally lowers the quality of the programming model.

Comment (JackAKirk) [reply sycl/plugins/unified_runtime/pi2ur.hpp:5512]: This is a key point a lot of people get confused over: cuContext has absolutely nothing to do with a sycl::context of a urContext. We map it 1:1 with cuDevice, matching CUDA runtime behaviour. It makes more sense to thing of cuContext as a device type in this sense which fits in with other backends like L0 that have ZeDevice as their "cuContext"

File: sycl/test-e2e/bindless_images/dx12_interop/read_write_unsampled.h
Comment (aelovikov-intel) [inline sycl/test-e2e/bindless_images/dx12_interop/read_write_unsampled.h:21]: Please use fine-grained includes, see https://github.com/intel/llvm/pull/13847.

File: sycl/include/sycl/detail/cg.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/cg.hpp:553]: Why not `std::optional`? Same in line 530.

File: sycl/plugins/unified_runtime/pi2ur.hpp
Comment (JackAKirk) [reply sycl/plugins/unified_runtime/pi2ur.hpp:5512]: > Am I missing something? Is there no good reason that almost every PI/UR function pairs the `device` parameter with a `context`?

If you can think of a good one apart from "It has always been like that", then you would be the first from what I know.

The only two existing reasons for sycl::context/ urContext are unrelated to any kind of usm based extension:

1. interop with opencl that has a context class for buffers: they could have just designed this to create a context for each opencl platform. The key point is this has nothing to do with your extension.
2. interop with opencl kernel bundles that also use the opencl context. The whole kernel_bundle thing also seems to be being redesigned. Again the most relevant point is this has nothing to do with your extension.

IMO both the above do not justify the "context" object being in a core programming language like SYCL, and causing a multiitude of problems for real programmers that we have to try to help them work around, but again this is besides the point. Both use cases have nothing to do with this extension.

Adding the concept to new extensions is just making the problem even bigger. Making programmers lives harder rather than easier.

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (aelovikov-intel) [inline sycl/plugins/cuda/pi_cuda.cpp:458]: When do we expect to drop that? Can/should we just wait until the ABI breaking window?

Also, I'm not even sure we need this deprecated entry point. Is it only needed for old libsycl.so that will use old plugin library?

File: sycl/plugins/unified_runtime/pi2ur.hpp
Comment (przemektmalon) [reply sycl/plugins/unified_runtime/pi2ur.hpp:5512]: I recognize that you make some good points on this issue. However, we've discussed this internally with the bindless images team, and we've decided to keep both `context` and `device` arguments in for now. Part of the reason is that we've done this for every API and we don't want to introduce API inconsistencies at this point. We also don't know whether the LevelZero implementation will require both a `context` and `device` to implement this functionality (some L0 `event` related APIs require a context, and optional device arguments, e.g. [zeEventPoolCreate](https://spec.oneapi.io/level-zero/latest/core/api.html#zeeventpoolcreate)).

To my knowledge importing of external semaphores/events is not yet supported in L0. At least I couldn't find anything related to it in the documentation, or any code samples. So I can't say whether L0 will need both device and context for this or not. But if creating event pools requires the context and optional device arguments, then again I'd rather be on the safe side and keep both. However, we will investigate what the semaphore/event interop API is supposed/planned to look like in LevelZero.

Bear in mind this is still an experimental extension. If we find out down the line that some or all of our APIs do not require both context and device arguments in every backend, then we're free to remove the context argument from our API calls throughout the entire stack. However, we can't commit to this large of a refactor at this point in time.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc
Comment (JackAKirk) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc:1726]: not: You could maybe write that opaque_fd isn't supported on windows.

Comment (przemektmalon) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc:1729]: I've added some minor notes to clarify that `opaque_fd` is only supported on linux based systems. I think going more into depth about which handle types are supported by which operating systems is beyond the scope of this document, and advanced users who want to use the interop APIs will have to know this already when exporting handles from third-party APIs before they even reach SYCL.

**Comment by przemektmalon** [comment]
Pinging @intel/bindless-images-reviewers @intel/dpcpp-nativecpu-pi-reviewers @intel/unified-runtime-reviewers

**Review by aelovikov-intel — APPROVED** [review]

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (JackAKirk) [inline sycl/plugins/unified_runtime/CMakeLists.txt:102]: Assuming you update this appropriately before merging. This LGTM.

**Review by JackAKirk — CHANGES_REQUESTED** [review]

**Review by PietroGhg — APPROVED** [review]
Native CPU LGTM, thank you

**Review by Seanst98 — APPROVED** [review]

**Comment by przemektmalon** [comment]
Friendly ping @intel/unified-runtime-reviewers

**Review by isaacault — APPROVED** [review]

**Review by DBDuncan — APPROVED** [review]

**Review by cppchedy — APPROVED** [review]

**Review by aelovikov-intel — APPROVED** [review]

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:130]:
```suggestion
  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
  # commit 269ba5430f2e6337c5ac25f9ef57b7e426e41fd0
  # Merge: 33eb5ea8 487f4f8a
  # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
  # Date:   Mon Jun 17 14:31:13 2024 +0100
  #     Merge pull request #1639 from przemektmalon/przemek/windows-interop
  #     [Bindless][Exp] Windows & DX12 interop. Semaphore ops can take values.
  set(UNIFIED_RUNTIME_TAG 269ba5430f2e6337c5ac25f9ef57b7e426e41fd0)
```

**Review by kbenzie — COMMENTED** [review]
https://github.com/oneapi-src/unified-runtime/pull/1639 has now been merged, please merge in the latest changes from the sycl branch, then update the UR repo/tag/comment as suggested.

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (kbenzie) [reply sycl/plugins/unified_runtime/CMakeLists.txt:102]: Thanks for the help policing the UR repo/tag @JackAKirk. I've seen various reviewers do similar so I recently added the *Unified Runtime Updates* secetion to the [Contribute to DPC++](https://intel.github.io/llvm-docs/developer/ContributeToDPCPP.html#unified-runtime-updates) page. Note that the [intel/unified-runtime-reviewers](https://github.com/orgs/intel/teams/unified-runtime-reviewers) team is responsible for not approving PR's unless the repo/tag are correct.

**Review by kbenzie — APPROVED** [review]

**Review by JackAKirk — APPROVED** [review]

**Comment by kbenzie** [comment]
@intel/llvm-gatekeepers please merge

**Pull request merged**
